src/urllib3/_collections.py

```diff
@@ -242,25 +341,39 @@ def extend(self, *args, **kwargs):
             for key, val in other.iteritems():
                 self.add(key, val)
         elif isinstance(other, Mapping):
-            for key in other:
-                self.add(key, other[key])
-        elif hasattr(other, "keys"):
-            for key in other.keys():
-                self.add(key, other[key])
-        else:
+            for key, val in other.items():
+                self.add(key, val)
+        elif isinstance(other, Iterable):
+            other = cast(Iterable[Tuple[str, str]], other)
             for key, value in other:
                 self.add(key, value)
+        elif hasattr(other, "keys") and hasattr(other, "__getitem__"):
+            # THIS IS NOT A TYPESAFE BRANCH
+            # In this branch, the object has a `keys` attr but is not a Mapping or any of
+            # the other types indicated in the method signature. We do some stuff with
+            # it as though it partially implements the Mapping interface, but we're not
+            # doing that stuff safely AT ALL.
+            for key in other.keys():
+                self.add(key, other[key])
 
         for key, value in kwargs.items():
             self.add(key, value)
 
-    def getlist(self, key, default=__marker):
+    @overload
+    def getlist(self, key: str) -> List[str]:
+        ...
+
+    @overload
+    def getlist(self, key: str, default: List[str]) -> List[str]:
```

Comment:
Wouldn't this override need to be `Union[List[str], DefaultType` where `DefaultType` is whatever type `default` is?

Reply:
Yeah, if we want to support some default that isn't `List[str]`, we'd need to do this:

```python3
@overload
def getlist(self, key: str, default: _DT) -> Union[List[str], _DT]]: ...
```
(in which case `Union[List[str], _DT]` collapses to `List[str]` in situations where `_DT` can be proven to be `List[str]`)

The question is whether we want to support passing a default that isn't `List[str]`; I can take a look through the existing codebase to see what we do and where; the most likely case is that we pass in `None` in some places, and at `typeshed` to see what they do for builtins that accept defaults.